File: Tutorials/ASP.NET/MVC5/CS/Controllers/CustomerController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using AspNetMvcApplication.Models;
using DevExpress.Xpo;
using XpoTutorial;

namespace AspNetMvcApplication.Controllers {

    public class CustomerController : BaseController {

        [ValidateInput(false)]
        public ActionResult CustomerListPartial() {
            var model = UnitOfWork.Query<Customer>()
                .Select(t => new CustomerViewModel() {
                    Oid = t.Oid,
                    FirstName = t.FirstName,
                    LastName = t.LastName
                });
            return PartialView("CustomerListPartial", model);
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult AddCustomer(CustomerViewModel model) {
            if(ModelState.IsValid) {
                SafeExecute(() => {
                    var customer = new Customer(UnitOfWork) {
                        FirstName = model.FirstName,
                        LastName = model.LastName
                    };
                    UnitOfWork.CommitChanges();
                });
            } else {
                ViewData["EditError"] = "Please, correct all errors.";
            }
            return CustomerListPartial();
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult UpdateCustomer(CustomerViewModel model) {
            if(ModelState.IsValid) {
                SafeExecute(() => {
                    var customer = UnitOfWork.GetObjectByKey<Customer>(model.Oid);
                    customer.FirstName = model.FirstName;
                    customer.LastName = model.LastName;
                    UnitOfWork.CommitChanges();
                });
            } else {
                ViewData["EditError"] = "Please, correct all errors.";
            }
            return CustomerListPartial();
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult DeleteCustomer(int Oid) {
            SafeExecute(() => {
                var customer = UnitOfWork.GetObjectByKey<Customer>(Oid);
                customer.Delete();
                UnitOfWork.CommitChanges();
            });
            return CustomerListPartial();
        }
    }
}

File: Tutorials/ASP.NET/MVC5/VB/Controllers/CustomerController.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.Mvc
Imports AspNetMvcApplication.Models
Imports DevExpress.Xpo
Imports XpoTutorial

Namespace AspNetMvcApplication.Controllers

    Public Class CustomerController
        Inherits BaseController

        <ValidateInput(False)>
        Public Function CustomerListPartial() As ActionResult
            Dim model = UnitOfWork.Query(Of Customer)().Select(Function(t) New CustomerViewModel() With {
                .Oid = t.Oid,
                .FirstName = t.FirstName,
                .LastName = t.LastName
            })
            Return PartialView("CustomerListPartial", model)
        End Function

        <HttpPost, ValidateInput(False)>
        Public Function AddCustomer(ByVal model As CustomerViewModel) As ActionResult
            If ModelState.IsValid Then
                SafeExecute(Sub()
                                Dim customer = New Customer(UnitOfWork) With {
                                    .FirstName = model.FirstName,
                                    .LastName = model.LastName
                                }
                                UnitOfWork.CommitChanges()
                            End Sub)
            Else
                ViewData("EditError") = "Please, correct all errors."
            End If
            Return CustomerListPartial()
        End Function

        <HttpPost, ValidateInput(False)>
        Public Function UpdateCustomer(ByVal model As CustomerViewModel) As ActionResult
            If ModelState.IsValid Then
                SafeExecute(Sub()
                                Dim customer = UnitOfWork.GetObjectByKey(Of Customer)(model.Oid)
                                customer.FirstName = model.FirstName
                                customer.LastName = model.LastName
                                UnitOfWork.CommitChanges()
                            End Sub)
            Else
                ViewData("EditError") = "Please, correct all errors."
            End If
            Return CustomerListPartial()
        End Function

        <HttpPost, ValidateInput(False)>
        Public Function DeleteCustomer(ByVal Oid As Integer) As ActionResult
            SafeExecute(Sub()
                            Dim customer = UnitOfWork.GetObjectByKey(Of Customer)(Oid)
                            customer.Delete()
                            UnitOfWork.CommitChanges()
                        End Sub)
            Return CustomerListPartial()
        End Function
    End Class
End Namespace